<?xml version="1.0" encoding="us-ascii"?>
<html><head><title>Running the SNP Viewer</title>
<meta content="text/html; charset=us-ascii" http-equiv="Content-Type" /><link rel="stylesheet" type="text/css" href="../Theme/nav.css" /><link rel="stylesheet" type="text/css" href="../Theme/popupmenu.css" /><link rel="stylesheet" type="text/css" href="../Theme/lf.css" /><link rel="stylesheet" type="text/css" href="../LinksExt/doc2help.css" /><link rel="stylesheet" type="text/css" href="../LinksExt/tranSMART_UserGuide.css" />
<script type="text/javascript">
//<![CDATA[



var _needIndentation = false;




//]]></script>

<script language="JavaScript" src="../linker.js" type="text/JavaScript"></script>

<script language="JavaScript" src="../special.js" type="text/JavaScript"></script>

<script language="JavaScript" src="../common.js" type="text/JavaScript"></script>

<meta name="Generator" content="ComponentOne Doc-To-Help v.2011.1.0.614" />
<meta name="D2HTheme" content="(Default) v.146" />
</head>
<body style="PADDING-BOTTOM: 0px; BACKGROUND-COLOR: #535353; MARGIN: 7px; PADDING-LEFT: 0px; PADDING-RIGHT: 0px; OVERFLOW: hidden; PADDING-TOP: 0px" onload="d2hOnLoadToolbar(event);" relpart="../">
<script language="javascript" type="text/javascript">
//<![CDATA[



d2hTopicPreOpen();



d2hTrySyncTOC();







var prev = true;





var next = true;



if (typeof d2hInitMainThemeHandlers != "undefined")

    d2hInitMainThemeHandlers(prev, next);

d2hAdjustTopicBorders(document);    



if (typeof d2hLoadWindows != "undefined")

    d2hLoadWindows();

if (typeof d2hLoadNavUrls != "undefined")

    d2hLoadNavUrls();




//]]></script>

<div style="PADDING-BOTTOM: 0px; BACKGROUND-COLOR: white; MARGIN: 0px; PADDING-LEFT: 0px; WIDTH: 100%; PADDING-RIGHT: 0px; HEIGHT: 100%; OVERFLOW: auto; PADDING-TOP: 0px" id="D2HTopicOuterBody" class="d2h_outer_topic">
<div id="nsbanner" class="d2hNavigatorTopic">
<div style="PADDING-BOTTOM: 0px; PADDING-LEFT: 0px; PADDING-RIGHT: 0px; PADDING-TOP: 0px" id="TitleRow" class="d2h_main_titlerow">
<div style="DISPLAY: none"><a id="D2HPreviousAnchor" href="generatingahaploview.htm"></a><a id="D2HNextAnchor" href="examples1.htm"></a></div></div>
<script language="javascript" type="text/javascript">
//<![CDATA[

if (typeof d2hProcessTopicNavForCSH != "undefined")

    d2hProcessTopicNavForCSH();


//]]></script>
</div>
<div style="PADDING-BOTTOM: 10px; PADDING-LEFT: 22px; PADDING-RIGHT: 10px; HEIGHT: 100%; PADDING-TOP: 5px" id="nstext" class="d2h_main_nstext" valign="bottom">
<div id="d2h_breadcrumbs" class="d2h_breadcrumbs"><span class="d2h_breadcrumbsLabel"><b><span style="COLOR: #000080">User's 
Guide</span></b>&#160;&#160;</span><a class="d2h_breadcrumbsNormal" href="chapter3datasetexplorer.htm">Chapter 3: Dataset Explorer</a></div>
<h2><a name="_Toc297057918" id="_Toc297057918">Running the SNP Viewer</a></h2>
<p class="MsoNormal">The SNP Viewer allows you to analyze individual base 
variations in DNA sequences for normal and tumor tissue samples in a SNP 
array.&#160; The viewer supports both copy number analysis and loss of 
heterozygosity (LOH) analysis.</p>
<p class="MsoNormal">SNP array data, such as Affymetrix Genome-Wide Human SNP 
Array, is required for the SNP Viewer. <span class="MsoHyperlink"><span style="FONT-SIZE: 9pt"></span></span></p>
<p class="MsoNormal">Dataset Explorer uses the Broad Institute&#8217;s GenePattern 
genomic analysis platform to generate SNP visualizations.&#160; For information 
about the Broad Institute&#8217;s SNPViewer, see the document <span class="Bold">SNPViewer </span>on the following site:</p>
<p class="MsoNormal"><a href="http://www.broadinstitute.org/cgi-bin/cancer/software/genepattern/modules/gp_modules.cgi"><span style="FONT-SIZE: 9pt">http://www.broadinstitute.org/cgi-bin/cancer/software/genepattern/modules/gp_modules.cgi</span></a><span class="MsoHyperlink"><span style="FONT-SIZE: 9pt"></span></span></p>
<p style="tab-stops: 0pt" class="ListNumStart">To run the SNP Viewer:</p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">1.&#160; Run tranSMART, then click the <span class="Bold">Dataset Explorer</span> tab.</p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">2.&#160; Select the study of interest.</p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">3.&#160; Drag the SNP array data into an empty box in one or 
both subsets.</p>
<p style="MARGIN-BOTTOM: 6pt; tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">4.&#160; Define the cohorts whose data points will be 
represented in the SNP visualization.</p>
<p style="MARGIN-BOTTOM: 6pt; tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">5.&#160; Click <span class="Bold">Advanced &gt; 
SNPViewer</span>.</p>
<p class="MsoNormalIndent">The SNPViewer dialog box appears:</p>
<p class="MsoNormalIndent"><img border="0" src="../ImagesExt/image17_158.jpg" alt="" /></p>
<p style="PAGE-BREAK-AFTER: avoid; MARGIN-BOTTOM: 6pt; tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">6.&#160; Select the data you would like to see in the 
visualization. Use one of the following filtering methods:</p>
<table style="BORDER-BOTTOM: medium none; BORDER-LEFT: medium none; BORDER-COLLAPSE: collapse; MARGIN-LEFT: 23.4pt; BORDER-TOP: medium none; BORDER-RIGHT: medium none" class="MsoTableGrid" border="1" cellspacing="0" cellpadding="0">
  <thead>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: windowtext 1.5pt double; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 121.5pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: black 1.5pt solid; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="162">
      <p class="TableHeading0">Filtering Method</p></td>
    <td style="BORDER-BOTTOM: windowtext 1.5pt double; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 297.9pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: black 1.5pt solid; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top" width="397">
      <p class="TableHeading0">Description</p></td></tr></thead>
  <tbody>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 121.5pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="162">
      <p class="TableText0">By Chromosome</p></td>
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 297.9pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top" width="397">
      <p class="TableText0">Select one or more chromosomes to include in the 
      visualization, or select <b>ALL</b> to include all chromosomes. </p>
      <p class="TableText0">To select multiple chromosomes, click a number, then 
      hold down <b>Ctrl</b> and click another.</p>
      <p class="TableText0"><b><span style="COLOR: #1f497d">Note: </span></b>If 
      gene or SNP rs IDs are entered, the selection of chromosomes is 
      ignored.<b><span style="COLOR: #1f497d"></span></b></p>
      <p class="TableText0"><b><span style="COLOR: #1f497d">Note: </span></b>For 
      performance reasons, only select chromosomes that are of particular 
      interest to you. </p></td></tr>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 121.5pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="162">
      <p class="TableText0">By Gene</p></td>
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 297.9pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top" width="397">
      <p class="TableText0">To select a gene, type all or part of the gene into 
      the <b>Add a Gene</b> field.</p>
      <p style="tab-stops: 12.0pt" class="TableBullet"><span style="POSITION: relative; FONT-FAMILY: 'Arial','sans-serif'; FONT-SIZE: 8pt; TOP: -2pt">&#9632;&#160;&#160; 
      </span>If you see the gene you want, click to select.</p>
      <p style="tab-stops: 12.0pt" class="TableBullet"><span style="POSITION: relative; FONT-FAMILY: 'Arial','sans-serif'; FONT-SIZE: 8pt; TOP: -2pt">&#9632;&#160;&#160; 
      </span>If you do not see the gene you want, type more characters.</p>
      <p class="TableText0">Separate any additional genes with commas.</p>
      <p class="TableText0"><span class="Bold"><span style="COLOR: #1f497d">Note:</span></span>&#160;&#160; Some genes that you 
      select as filters in tranSMART may not be found in the GenePattern 
      SNPViewer.</p></td></tr>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 121.5pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="162">
      <p class="TableText0">By SNP rs ID</p></td>
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 297.9pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top" width="397">
      <p class="TableText0">To add a SNP rs ID, type the full ID into the 
      <b>Selected SNPs</b> field. Separate any additional IDs with 
  commas.</p></td></tr>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: black 1.5pt solid; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 121.5pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="162">
      <p class="TableText0">By a combination of genes or SNP rs IDs</p></td>
    <td style="BORDER-BOTTOM: black 1.5pt solid; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 297.9pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top" width="397">
      <p class="TableText0">Follow the instructions above to add a combination of 
      genes or SNP rs IDs.&#160; </p></td></tr></tbody></table>
<p style="TEXT-INDENT: 0pt" class="MsoListNumber">&#160;</p>
<p style="PAGE-BREAK-AFTER: avoid; tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">7.&#160; Click <span class="Bold">OK</span>.</p>
<p style="PAGE-BREAK-AFTER: avoid; MARGIN-BOTTOM: 6pt" class="MsoNormalIndent"><span style="COLOR: black">If any security dialog boxes 
appear, acknowledge them by clicking <b>Continue</b> or <b>OK</b>. </span></p>
<p style="MARGIN-BOTTOM: 6pt" class="MsoNormalIndent"><span style="COLOR: black">The visualization may require several minutes to display a 
large dataset, such as when multiple chromosomes are selected.</span></p>
<hr size="1" noshade="" />

<p class="RelatedHead">
</p><p style="FONT-FAMILY: arial; COLOR: #000080">Related Topics</p>
<p></p>
<p class="MsoNormal"><a style="TEXT-DECORATION: none" href="examples1.htm"><img border="0" alt="" align="absMiddle" src="../button.gif" />Examples</a></p>
<p class="MsoNormal"><a style="TEXT-DECORATION: none" href="viewingsnpdata.htm"><img border="0" alt="" align="absMiddle" src="../button.gif" />Viewing SNP Data</a></p></div></div></body></html>
